pallets/oracle-data-collection/src/tests.rs

```diff
@@ -0,0 +1,309 @@
+use cfg_traits::data::DataRegistry;
+use frame_support::{assert_err, assert_ok, storage::bounded_vec::BoundedVec};
+use sp_runtime::{testing::H256, traits::Get, DispatchError};
+
+use crate::{
+	mock::*,
+	pallet::{Config, Error, Keys},
+	types::Change,
+};
+
+const ADMIN: AccountId = 1;
+const ANY: AccountId = 100;
+const FEEDER_1: AccountId = 10;
+const FEEDER_2: AccountId = 11;
+const FEEDER_3: AccountId = 12;
+
+const COLLECTION_ID: CollectionId = 1;
+const KEY_A: OracleKey = 1;
+const KEY_B: OracleKey = 2;
+const KEY_ERR: OracleKey = 3;
+const KEY_NONE: OracleKey = 4;
+const CHANGE_ID: ChangeId = H256::repeat_byte(0x42);
+
+mod mock {
+	use super::*;
+
+	pub fn prepare_update_feeders(
+		key: OracleKey,
+		feeders: &BoundedVec<AccountId, MaxFeedersPerKey>,
+	) {
+		MockChangeGuard::mock_note({
+			let feeders = feeders.clone();
+			move |pool_id, change| {
+				assert_eq!(pool_id, COLLECTION_ID);
+				assert_eq!(change, Change::Feeders(key, feeders.clone()));
+				Ok(CHANGE_ID)
+			}
+		});
+		MockChangeGuard::mock_released({
+			let feeders = feeders.clone();
+			move |pool_id, change_id| {
+				assert_eq!(pool_id, COLLECTION_ID);
+				assert_eq!(change_id, CHANGE_ID);
+				Ok(Change::Feeders(key, feeders.clone()))
+			}
+		});
+		MockIsAdmin::mock_check(|(admin, collection_id)| {
+			assert_eq!(admin, ADMIN);
+			assert_eq!(collection_id, COLLECTION_ID);
+			true
+		});
+	}
+
+	pub fn prepare_provider() {
+		MockProvider::mock_get(|(account, collection_id), key| {
+			assert_eq!(collection_id, &COLLECTION_ID);
+			match (account, key) {
+				(&FEEDER_1, &KEY_A) => Ok(Some((100, 50))),
+				(&FEEDER_2, &KEY_A) => Ok(Some((101, 45))),
+				(&FEEDER_3, &KEY_A) => Ok(Some((102, 55))),
+				(&FEEDER_1, &KEY_B) => Ok(Some((1000, 500))),
+				(&FEEDER_2, &KEY_B) => Ok(None),
+				(&FEEDER_3, &KEY_B) => Ok(None),
+				(&FEEDER_1, &KEY_ERR) => Err(DispatchError::Other("get err")),
+				(&FEEDER_2, &KEY_ERR) => Err(DispatchError::Other("get err")),
+				(&FEEDER_3, &KEY_ERR) => Err(DispatchError::Other("get err")),
+				(&FEEDER_1, &KEY_NONE) => Ok(None),
+				(&FEEDER_2, &KEY_NONE) => Ok(None),
+				(&FEEDER_3, &KEY_NONE) => Ok(None),
+				_ => unreachable!(),
+			}
+		});
+	}
+}
+
+mod util {
+	use super::*;
+
+	pub fn update_feeders(key: OracleKey, feeders: Vec<AccountId>) {
+		let feeders: BoundedVec<_, _> = feeders.try_into().unwrap();
+
+		MockChangeGuard::mock_note(|_, _| Ok(CHANGE_ID));
+		MockChangeGuard::mock_released({
+			let feeders = feeders.clone();
+			move |_, _| Ok(Change::Feeders(key, feeders.clone()))
+		});
+		MockIsAdmin::mock_check(|_| true);
+
+		OracleCollection::propose_update_feeders(
+			RuntimeOrigin::signed(ADMIN),
+			COLLECTION_ID,
+			key,
+			feeders,
+		)
+		.unwrap();
+
+		OracleCollection::apply_update_feeders(
+			RuntimeOrigin::signed(ADMIN),
+			COLLECTION_ID,
+			CHANGE_ID,
+		)
+		.unwrap();
```

Comment:
Nit: Would be cleaner to wrap the calls into `assert_ok!(..)`

Reply:
Yeah! I have a nitpick reason for this. 

I put `assert_ok!` in lines that the test should evaluate and `unwrap()` in lines where it's expected to work but does not belong to the evaluation of the test itself.

In this case, those lines are under `util::update_feeders()`, which is a utility whose concern is not to evaluate if it works fine but to provide some state to later use in other tests.

The well behaviour of those calls with `assert_ok!` is done in `updating_feeders()`

P.S. Quite nitpick probalbly 😆

Reply:
Makes sense! Thanks for the explanation :)